src/main/java/org/galatea/starter/service/IexService.java

```diff
@@ -47,22 +69,134 @@
+        for (IexHistoricalPrice iexPrice:historicalPrice)
+          historicalPriceRpsy.save(iexPrice);
+        return historicalPrice;
+      }
+    }
+  }
+
+  private List<IexHistoricalPrice> getHistoricalPricesForSymbolByRange(final String symbol, final String range, final String token) {
+    List<IexHistoricalPrice> historicalPrice = new ArrayList<>();
+    LocalDate currentDate = LocalDate.now();
+    LocalDate startingDate = null;
+    List<LocalDate> dates = null;
+    List<IexHistoricalPrice> prices = null;
+
+    if (range == null || range.equals(""))
+      startingDate = checkStartingDate(currentDate, RANGE_DEFAULT);
+    else
+      startingDate = checkStartingDate(currentDate, range);
+
+    if (startingDate == null)
+      throw new IllegalArgumentException("Invalid range over date: Null pointer for date.");
+    log.info("Logging: Today = " + currentDate.toString() + " startingDate = " + startingDate.toString());
+    dates = startingDate.datesUntil(currentDate, Period.ofDays(1)).collect(Collectors.toList());
+    prices = getHistoricalPricesForSymbolCache(symbol, dates);
+    if (!prices.isEmpty())
+      return prices;
     else
-      if(date == null || date.equals(""))
+      return getHistoricalPricesForSymbolIEX(symbol, range, token);
+  }
+
+  private LocalDate checkStartingDate(LocalDate today, String range){
+    if (range.equals(RANGE_YEAR_TO_DATE))
+      return LocalDate.of(today.getYear(), 1, 1);
+    else if (range.equals(RANGE_MAX))
+      return today.minusYears(MAX_YEAR);
+
+    String[] range_part = range.split("(?<=\\d)(?=\\D)");
+    int time_num = 0;
+    String time_span = null;
+    try {
+      time_num= Integer.parseInt(range_part[0]);
+      time_span = range_part[1];
+    }
+    catch (Exception e) {
+      log.info(e.toString());
+    }
+
+    LocalDate startingDate;
+    switch (time_span) {
+      case RANGE_DAY:
+        startingDate = today.minusDays(time_num);
+        break;
+      case RANGE_MONTH:
+        startingDate = today.minusMonths(time_num);
+        break;
+      case RANGE_YEAR:
+        startingDate = today.minusYears(time_num);
+        break;
+      default:
+        startingDate = null;
+        break;
+    }
+    return startingDate;
+  }
+
+  public List<IexHistoricalPrice> getHistoricalPricesForSymbolCache(final String symbol, final List<LocalDate> dates) {
+    List<IexHistoricalPrice> historicalPrice = new ArrayList<>();
+    log.info("Logging: All data from cache "+ historicalPriceRpsy.findAll().toString());
+    for (LocalDate date:dates) {
+      if (date.getDayOfWeek() == DayOfWeek.SATURDAY || date.getDayOfWeek() == DayOfWeek.SUNDAY || holidays.getHolidays().contains(date))
+        continue;
+      IexPriceID iexPriceID = new IexPriceID(symbol.toUpperCase(), date.toString());
+
+      Optional<IexHistoricalPrice> price = historicalPriceRpsy.findById(iexPriceID);
+      if (price.isPresent()) {
+        historicalPrice.add(price.get());
```

Comment:
Generally when dealing with optionals, you want to avoid using .get
try price.ifPresent(historicalPrice::add);